### PR TITLE
Exploration of technical challenges involved with adding Kokkos support to RRTMGP

### PR DESCRIPTION
**Pull request by jgfouca**
We (eamxx) are interested in RRTMGP having the ability to use Kokkos. This would reduce the need for data copying in the interface layer and reduce some of our devops costs by eliminating the YAKL dependency.

I thought it would be best to spend a little bit of time "in the trench" before coming up with a proposal, so I spent the last day or so starting the first phase of the conversion and I made some progress which became this PR. Kokkos support was added to the free function read_atmos and the class GasConcs.

### Phase 1

The focus will be on rrtmgp_allsky.cpp test, which appears to be the main test program for rrtmgp. I have made some minor upgrades to the `cpp/test/build` area to make things a bit easier to build and test.

The phase1 approach will be a "dumb" conversion where all classes and functions that do kernels or contain yakl arrays will be fully duplicated. The duplicated version will be switched to kokkos and have a member function `validate_kokkos` added which takes an object from the original class and checks that all the members are equivalent. It's up the caller, the rrtmgp_allsky test in this case, to invoke `validate_kokkos`.

### Phase 2

Performance checking between Kokkos and YAKL versions. Kokkos version may need some optimizations, especially with temporary arrays (need WorkspaceManager?).

### Phase 3 (optional)

Once a critical mass of code has been ported, we can begin to look at a better design that removes duplication and uses software design to make the code more agnostic as the what kernel launcher is being used, maybe resorting to ifdefs for the pieces that are hard to make agnostic. Effort here depends on how much we think we will want to keep YAKL support longterm. 

### Progress update

Phase 1 is complete.

### CODE REVIEW

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -17,6 +17,9 @@ set(CXX_SRC
 # Add RRTMGP as a library and make sure we link against YAKL
 add_library(rrtmgp STATIC ${CXX_SRC})
 target_link_libraries(rrtmgp yakl)
+if (RRTMGP_ENABLE_KOKKOS)
```
Comment (bartgol) [inline cpp/CMakeLists.txt:20]: For the sake of rrtmgp as a standalone package, maybe we should expose this var as a CACHE var in the main CMakeLists.txt, and also add a `find_package` call. That way, if some other package wants to use rrtmgp with kokkos, they at least have a somewhat well defined (in a CMake sense) way of triggering it.

File: cpp/conversion.h [hunk 2]
```diff
@@ -0,0 +1,706 @@
+#pragma once
```
Comment (bartgol) [inline cpp/conversion.h]: For the sake of rrtmgp integration in other code sources, this file could prob have a more unique name. Something like `rrtmgp_kokkos_integration_utils.hpp` or simply `rrtmgp_kokkos_conversion.h`.

File: cpp/conversion.h [hunk 3]
```diff
@@ -0,0 +1,706 @@
...
+    throw exception_type(_ss_.str());                     \
+  }                                                       \
+} while(0)
+
+#define RRT_REQUIRE(condition, msg) IMPL_THROW_RRT(condition, msg, std::runtime_error)
+
+// Copied from EKAT
+template<typename T, int N>
+struct DataND {
+  using type = typename DataND<T,N-1>::type*;
```
Comment (bartgol) [inline cpp/conversion.h]: I can foresee some typos or other code errors that cause the instantiation of DataND with N<0...

If you added a `static_assert(N>=0, "SOME GOOD ERROR MSG");`, the compiler will stop immediately and give a clear error msg, rather than waste time on recursion and report some generic recursion limit error.

File: cpp/conversion.h (hunk 2 above)
Comment (bartgol) [reply cpp/conversion.h]: Also, minor, but since this is a C++ file, you should probably use `.hpp`. The `.h` file may give the false feeling that it's ok to include it in a C source code..

File: cpp/examples/all-sky/mo_garand_atmos_io.cpp [hunk 4]
```diff
@@ -94,7 +89,105 @@ void read_atmos(std::string input_file, real2d &p_lay, real2d &t_lay, real2d &p_
...
+
+  p_lay = real2dk("p_lay",ncol,nlay);
+  t_lay = real2dk("t_lay",ncol,nlay);
+  p_lev = real2dk("p_lev",ncol,nlev);
+  t_lev = real2dk("t_lev",ncol,nlev);
+
+  real2dk tmp2d;
+  // p_lay
+  io.read(tmp2d,"p_lay");
+  parallel_for( ncol, KOKKOS_LAMBDA (int icol) {
```
Comment (bartgol) [inline cpp/examples/all-sky/mo_garand_atmos_io.cpp]: I wonder if certain compilers (perhaps in particular situations) may not understand which ||for you want: yakl's or kokkos's. It is possible the user may issue something like `using yakl::c::parallel_for;` somewhere in a previous header.

Perhaps it's best to keep things explicit, and use `Kokkos::parallel_for`.It's only few extra chars.

File: cpp/extensions/cloud_optics/mo_cloud_optics.h [hunk 5]
```diff
@@ -62,20 +62,7 @@ class CloudOptics : public OpticalProps {
...
-    real radliq_upr = 0;
-    real radice_lwr = 0;
-    real radice_upr = 0;
-    int liq_nsteps = 0;
-    int ice_nsteps = 0;
-    real liq_step_size = 0;
-    real ice_step_size = 0;
-  }
-
+  CloudOptics() = default;
```
Comment (bartgol) [inline cpp/extensions/cloud_optics/mo_cloud_optics.h:65]: Keep in mind that default constructor does _not_ zero-initialize. If zero-initialization is important, you need to revert this change or add `= 0;` next to the member declaration.

**Review by bartgol — APPROVED** [review]
A lot of work in a very short time! Overall I think this is good, but I have some thoughts.

File: cpp/conversion.h (hunk 2 above)
Comment (jgfouca) [reply cpp/conversion.h]: Agreed on both. The pattern in this codebase is for all headers to be `.h` unfortunately, so I just continued with that.

File: cpp/CMakeLists.txt (hunk 1 above)
Comment (jgfouca) [reply cpp/CMakeLists.txt:20]: The cmake situation in rrtmgp is a little weird. `cpp/test` is a subdirectory of `cpp` but the CMakeLists.txt in `test` is the one that sets up the project and adds YAKL and Kokkos. It's clear that `cpp/CMakeLists.txt` only expects to be used by an external project that has already set up YAKL and Kokkos.

File: cpp/conversion.h (hunk 3 above)
Comment (jgfouca) [reply cpp/conversion.h]: I think we may just use the EKAT version of this once the conversion is done.

File: cpp/examples/all-sky/mo_garand_atmos_io.cpp (hunk 4 above)
Comment (jgfouca) [reply cpp/examples/all-sky/mo_garand_atmos_io.cpp]: Agree.

File: cpp/extensions/cloud_optics/mo_cloud_optics.h (hunk 5 above)
Comment (jgfouca) [reply cpp/extensions/cloud_optics/mo_cloud_optics.h:65]: Couple things. First, notice these are all local variable decls, not members. I think the coder intended to change members, since otherwise it makes no sense to do this, but my change doesn't actually change how the code behaves. Second, are you sure this will not zero initialize the members? Doesn't the default constructor promise to call `()` on all members, which for primitives, would zero initialize?

Comment (jgfouca) [reply cpp/extensions/cloud_optics/mo_cloud_optics.h:65]: Also, my valgrind runs were not reporting uninitialized memory use.

Comment (bartgol) [reply cpp/extensions/cloud_optics/mo_cloud_optics.h:65]: Oh, you're right. I didn't read carefully the code. Then yes, what you did is perfect.

Comment (mrnorman) [reply cpp/extensions/cloud_optics/mo_cloud_optics.h:65]: I was likely the original coder, and I agree, that constructor is useless and certainly a mistake on my part. Thanks for correcting that.

File: cpp/examples/all-sky/mo_garand_atmos_io.cpp (hunk 4 above)
Comment (mrnorman) [reply cpp/examples/all-sky/mo_garand_atmos_io.cpp]: I agree as well. We should have local `using` statements only or explicit namespacing everywhere.

Comment (mrnorman) [reply cpp/examples/all-sky/mo_garand_atmos_io.cpp]: I swear I've gotten better at C++ over time...

**Comment by mrnorman** [comment]
I'm curious: Are you going to insert Packs into RRTMGP++? I think that'd be required in order to reduce data copying on CPUs. If you need clarification on anything with YAKL use, don't hesitate to reach out.

File: cpp/rrtmgp_const.h [hunk 6]
```diff
@@ -81,11 +94,76 @@ typedef FArray<char,2,yakl::memHost> charHost2d;
 
 typedef FArray<std::string,1,yakl::memHost> string1d;
 
+#ifdef RRTMGP_ENABLE_KOKKOS
+typedef FView<real*>       real1dk;
+typedef FView<real**>      real2dk;
+typedef FView<real***>     real3dk;
+typedef FView<real****>    real4dk;
+typedef FView<real*****>   real5dk;
```
Comment (mrnorman) [inline cpp/rrtmgp_const.h:113]: @bartgol , It looks like you're leaving the index ordering the same in the Kokkos kernels. If that's the case, then shouldn't these be `LayoutLeft`? I'm curious how you got the same results without either `LayoutLeft` or permuting the index ordering.

File: cpp/examples/all-sky/mo_garand_atmos_io.cpp [hunk 7]
```diff
@@ -50,23 +51,16 @@ void read_atmos(std::string input_file, real2d &p_lay, real2d &t_lay, real2d &p_
...
-  string1d gas_names("gas_names",ngas);
-  gas_names(1) = std::string("h2o");
-  gas_names(2) = std::string("co2");
-  gas_names(3) = std::string("o3" );
-  gas_names(4) = std::string("n2o");
-  gas_names(5) = std::string("co" );
-  gas_names(6) = std::string("ch4");
-  gas_names(7) = std::string("o2" );
-  gas_names(8) = std::string("n2" );
+  std::vector<std::string> gas_names = {
```
Comment (mrnorman) [inline cpp/examples/all-sky/mo_garand_atmos_io.cpp:52]: Thanks for doing this. Using `string1d` was kinda dumb.

File: cpp/rrtmgp/mo_gas_concentrations.h [hunk 8]
```diff
@@ -102,6 +115,15 @@ class GasConcs {
     parallel_for( YAKL_AUTO_LABEL() , SimpleBounds<2>(nlay,ncol) , YAKL_LAMBDA (int ilay, int icol) {
       this_concs(icol,ilay,igas) = w;
     });
+#ifdef RRTMGP_ENABLE_KOKKOS
+    auto this_concs_k = this->concs_k;
+    Kokkos::parallel_for(nlay, KOKKOS_LAMBDA(int ilay) {
+      for (int icol = 0; icol < ncol; ++icol) {
```
Comment (mrnorman) [inline cpp/rrtmgp/mo_gas_concentrations.h]: I'm curious why ncol is being serialized.

File: cpp/rrtmgp_const.h (hunk 6 above)
Comment (jgfouca) [reply cpp/rrtmgp_const.h:113]: @mrnorman , see above:
```
using FView = Kokkos::View<T, Kokkos::LayoutLeft, Device>;
```

All these FViews are layout left.

Comment (mrnorman) [reply cpp/rrtmgp_const.h:113]: Ahh, Sorry, I missed that. Thanks.

Comment (bartgol) [reply cpp/rrtmgp_const.h:113]: I think these _are_ `LayoutLeft` views. The F in front of the type name makes me thnk "FortranView", which probably means `LayoutLeft`. @jgfouca may correct me though.

File: cpp/rrtmgp/mo_gas_concentrations.h (hunk 8 above)
Comment (jgfouca) [reply cpp/rrtmgp/mo_gas_concentrations.h]: @mrnorman , that's a good question. In YAKL, when you have something like this:
```
parallel_for( YAKL_AUTO_LABEL() , SimpleBounds<2>(nlay,ncol), YAKL_LAMBDA (int ilay, int icol) {
   ...
```

Does that parallelize across both indices? I assumed it was only parallelizing on nlay. In Kokkos, parallelizing on both would require hierarchical parallelism which is a big step up in complexity since you need to set up thread teams for that.

Comment (bartgol) [reply cpp/rrtmgp/mo_gas_concentrations.h]: @jgfouca adding to Matt's comment: if the view is LayoutLeft, then we can take a subview `subview(ALL,ALL,igas-1)`, and then just do `deep_copy(sv,w);`.

**Comment by jgfouca** [comment]
@mrnorman , thanks for chiming in! Your input is especially valuable here.

> I'm curious: Are you going to insert Packs into RRTMGP++? I think that'd be required in order to reduce data copying on CPUs. If you need clarification on anything with YAKL use, don't hesitate to reach out.

Yes, I think we will likely integrate RRTMGP with EKAT or use templating so view's of packs are supported.

File: cpp/conversion.h (hunk 2 above)
Comment (mrnorman) [reply cpp/conversion.h]: You can feel free to change it to `.hpp`. I don't mind that.

File: cpp/rrtmgp/mo_gas_concentrations.h (hunk 8 above)
Comment (bartgol) [reply cpp/rrtmgp/mo_gas_concentrations.h]: Jim, you could also use MDRangePolicy, which I think it's the kokkos equivalent of SimpleBounds<N> in yakl.

Comment (mrnorman) [reply cpp/rrtmgp/mo_gas_concentrations.h]: > @mrnorman , that's a good question. In YAKL, when you have something like this:
> 
> ```
> parallel_for( YAKL_AUTO_LABEL() , SimpleBounds<2>(nlay,ncol), YAKL_LAMBDA (int ilay, int icol) {
>    ...
> ```
> 
> Does that parallelize across both indices? I assumed it was only parallelizing on nlay. In Kokkos, parallelizing on both would require hierarchical parallelism which is a big step up in complexity since you need to set up thread teams for that.

YAKL parallelizes all dimensions added to the `Bounds` or `SimpleBounds` class in order from left to right being outermost to innermost in the threading / implied looping. More information here:

https://github.com/mrnorman/YAKL/wiki/parallel_for-and-Bounds-(C-and-Fortran-Styles),-and-synchronization

Comment (mrnorman) [reply cpp/rrtmgp/mo_gas_concentrations.h]: Right, RangePolicy is roughly equivalent to a 1-D Bounds object, and MDRangePolicy is roughly equivalent to a multi-D Bounds object.

Comment (jgfouca) [reply cpp/rrtmgp/mo_gas_concentrations.h]: Thanks guys, I will switch this pattern to MDRangePolicy

Comment (mrnorman) [reply cpp/rrtmgp/mo_gas_concentrations.h]: Regarding your comment about hierarchical parallelism, YAKL can support multiple **tightly nested** loops by essentially collapsing them, which I assume is what `MDRangePolicy` does in its backend as well. YAKL does have two-level hierarchical parallelism, but it's not needed for this code.

Comment (jgfouca) [reply cpp/rrtmgp/mo_gas_concentrations.h]: Thanks @mrnorman , that will simplify things a lot!

File: cpp/rrtmgp/mo_gas_concentrations.h [hunk 9]
```diff
@@ -132,8 +154,51 @@ class GasConcs {
...
+
+    if (w.extent(0) != this->nlay) { stoprun("GasConcs::set_vmr: different dimension (nlay)"); }
+    int igas = this->find_gas(gas);
+    if (igas == GAS_NOT_IN_LIST) {
+      stoprun("GasConcs::set_vmr(): trying to set a gas whose name not provided at initialization");
+    }
+    // Check for bad values in w
+    #ifdef RRTMGP_EXPENSIVE_CHECKS
+    bool badVal = false;
+    Kokkos::parallel_reduce(w.extent(0), KOKKOS_LAMBDA(int i, bool& is_bad) {
```
Comment (tcclevenger) [inline cpp/rrtmgp/mo_gas_concentrations.h]: Will using `w.extent(0)` be a problem when we switch to packs? Maybe use `nlay` like you do below.

Comment (tcclevenger) [reply cpp/rrtmgp/mo_gas_concentrations.h]: Actually, I see the assert above, so that will catch any issues with packs when that part of the conversion comes. So ignore.

**Comment by jgfouca** [comment]
Phase 1 is complete so this is ready for final review and merge. I don't think there should be any impacts to users if kokkos is not enabled.

File: cpp/examples/all-sky/rrtmgp_allsky.cpp [hunk 10]
```diff
@@ -147,12 +227,33 @@ int main(int argc , char **argv) {
...
+      Kokkos::parallel_for( MDRangeP<2>({0,0}, {nlay,ncol}) , KOKKOS_LAMBDA (int ilay, int icol) {
+        cloud_mask_k(icol,ilay) = p_lay_k(icol,ilay) > 100. * 100. && p_lay_k(icol,ilay) < 900. * 100. && mod(icol+1, 3) != 0;
+        // Ice and liquid will overlap in a few layers
+        lwp_k(icol,ilay) = merge(10.,  0., cloud_mask_k(icol,ilay) && t_lay_k(icol,ilay) > 263.);
+        iwp_k(icol,ilay) = merge(10.,  0., cloud_mask_k(icol,ilay) && t_lay_k(icol,ilay) < 273.);
+        rel_k(icol,ilay) = merge(rel_val, 0., lwp_k(icol,ilay) > 0.);
+        rei_k(icol,ilay) = merge(rei_val, 0., iwp_k(icol,ilay) > 0.);
+      });
+      {
+        conv::compare_yakl_to_kokkos(cloud_mask, cloud_mask_k);
```
Comment (brhillman) [inline cpp/examples/all-sky/rrtmgp_allsky.cpp:240]: Why does cloud_mask not get compared in the below call to compare_all_yakl_to_kokkos?

File: cpp/rrtmgp_conversion.h [hunk 11]
```diff
@@ -0,0 +1,956 @@
...
+do {                                                      \
+  if ( ! (condition) ) {                                  \
+    std::stringstream _ss_;                               \
+    _ss_ << "\n FAIL:\n" << #condition  << "\n";          \
+    _ss_ << "\n" << msg;                                  \
+    throw exception_type(_ss_.str());                     \
+  }                                                       \
+} while(0)
+
+#define RRT_REQUIRE(condition, msg) IMPL_THROW_RRT(condition, msg, std::runtime_error)
```
Comment (brhillman) [inline cpp/rrtmgp_conversion.h:136]: In general it would be nice to have a little more documentation in this header file for the functions it implements to explain what everything is doing. Most are self explanatory enough though I guess.

**Comment by brhillman** [comment]
This looks great, thanks Jim!

**Review by brhillman — APPROVED** [review]
Left a few comments, but nothing critical.

**Review by bartgol — APPROVED** [review]
I did not look at the whole code. Since this is mostly a yakl->kokkos translation via code duplication, and I think you tested that the answer is unchanged, I think this is fine.

That said, I strongly encourage a follow-up PR, where the two impl's are merged into one via typedefs and macros, otherwise the maintenance cost will be very high.

**Review by tcclevenger — APPROVED** [review]
Looks good, Jim!

File: cpp/examples/all-sky/rrtmgp_allsky.cpp (hunk 10 above)
Comment (jgfouca) [reply cpp/examples/all-sky/rrtmgp_allsky.cpp:240]: Because it's a different type, an array of bools, not reals.

File: cpp/rrtmgp_conversion.h (hunk 11 above)
Comment (jgfouca) [reply cpp/rrtmgp_conversion.h:136]: Good point, I will add some.
